### PR TITLE
daml start: rerun the initialization script on rebuild

### PR DESCRIPTION
**Pull request by robin-da**
We rerun the initialization script for every rebuild. In addition we fix
a potential bug that would lead to a wrong sandbox port when trying to
re-upload the package.

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 1]
```diff
@@ -228,26 +228,29 @@ runStart
...
+                      , if any (`elem` ["-s", "--static-time"]) sandboxOpts
+                          then "--static-time"
+                          else "--wall-clock-time"
+                      , "--ledger-host"
+                      , "localhost"
+                      , "--ledger-port"
+                      , case sandboxPort of SandboxPort port -> show port
+                      ] ++ scriptOpts
+                  runProcess_ procScript
+        doRunInitScript
```
Comment (cocreature) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:251]: We now run the init script before navigator is up. I guess there is no reason not to do this?

**Review by cocreature — APPROVED** [review]
Nice!

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs (hunk 1 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:251]: I couldn't think of any. In fact, I think it's safer to make sure the navigator reflects the most up to date state.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 2]
```diff
@@ -286,15 +289,16 @@ runStart
...
+        rebuild :: ProjectConfig -> FilePath -> SandboxPort -> IO () -> IO ()
+        rebuild projectConfig darPath sandboxPort doRunInitScript = do
+          putStrLn "Re-building and uploading package."
           doBuild
           doCodegen projectConfig
-          doUploadDar darPath
-          putStrLn "rebuild complete"
+          doUploadDar darPath sandboxPort
+          doRunInitScript
+          putStrLn "Rebuild complete."
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:301]:
```suggestion
          putStrLn "Rebuild and upload complete."
```

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 3]
```diff
@@ -286,15 +289,16 @@ runStart
...
                 c <- getChar
-                when (c == 'r' || c == 'R') $ rebuild projectConfig darPath
+                when (c == 'r' || c == 'R') $ rebuild projectConfig darPath sandboxPort runInitScript
                 threadDelay 1000000
-
-        rebuild projectConfig darPath = do
-          putStrLn "re-building and uploading package"
+        rebuild :: ProjectConfig -> FilePath -> SandboxPort -> IO () -> IO ()
+        rebuild projectConfig darPath sandboxPort doRunInitScript = do
+          putStrLn "Re-building and uploading package."
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:296]: I like it to end such lines with `...` to signal that something is going on. But's that obviously a matter of personal taste.

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs [hunk 4]
```diff
@@ -228,26 +228,29 @@ runStart
...
+                      , initScript
+                      , if any (`elem` ["-s", "--static-time"]) sandboxOpts
+                          then "--static-time"
+                          else "--wall-clock-time"
+                      , "--ledger-host"
+                      , "localhost"
+                      , "--ledger-port"
+                      , case sandboxPort of SandboxPort port -> show port
+                      ] ++ scriptOpts
+                  runProcess_ procScript
```
Comment (hurryabit) [inline daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:250]: This is getting more and more complex. Do you think there's a chance we could have an integration test that the `R` keypress works as intended?

**Review by hurryabit — APPROVED** [review]
Awesome. Thanks a lot!

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs (hunk 4 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:250]: Yes we should probably add one. Would you be ok to do this in a follow-up PR as I think that test will be a bit more involved?

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs (hunk 3 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs]: added some dots ...

File: daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs (hunk 4 above)
Comment (robin-da) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:250]: I merge this one for now and start working on an integration test.

Comment (hurryabit) [reply daml-assistant/daml-helper/src/DA/Daml/Helper/Start.hs:250]: Adding this in a separate PR would have been my suggestion as well. Thanks for taking care of it.

**Pull request merged**